invokeai/app/services/shared/invocation_context.py

```diff
@@ -0,0 +1,625 @@
+            :param is_intermediate: Whether the image is an intermediate. Intermediate images aren't added to the gallery.
+            """
+            if is_intermediate is not None:
+                services.images.update(image_name, ImageRecordChanges(is_intermediate=is_intermediate))
+            if board_id is None:
+                services.board_images.remove_image_from_board(image_name)
+            else:
+                services.board_images.add_image_to_board(image_name, board_id)
+            return services.images.get_dto(image_name)
+
+        self.save = save
+        self.get_pil = get_pil
+        self.get_metadata = get_metadata
+        self.get_dto = get_dto
+        self.update = update
+
+
+class LatentsInterface:
+    def __init__(
+        self,
+        services: InvocationServices,
+        context_data: InvocationContextData,
+    ) -> None:
+        def save(tensor: Tensor) -> str:
+            """
+            Saves a latents tensor, returning its name.
+
+            :param tensor: The latents tensor to save.
+            """
+
+            # Previously, we added a suffix indicating the type of Tensor we were saving, e.g.
+            # "mask", "noise", "masked_latents", etc.
+            #
+            # Retaining that capability in this wrapper would require either many different methods
+            # to save latents, or extra args for this method. Instead of complicating the API, we
+            # will use the same naming scheme for all latents.
+            #
+            # This has a very minor impact as we don't use them after a session completes.
+
+            # Previously, invocations chose the name for their latents. This is a bit risky, so we
+            # will generate a name for them instead. We use a uuid to ensure the name is unique.
+            #
+            # Because the name of the latents file will includes the session and invocation IDs,
+            # we don't need to worry about collisions. A truncated UUIDv4 is fine.
+
+            name = f"{context_data.session_id}__{context_data.invocation.id}__{uuid_string()[:7]}"
+            services.latents.save(
+                name=name,
+                data=tensor,
+            )
+            return name
+
+        def get(latents_name: str) -> Tensor:
+            """
+            Gets a latents tensor by name.
+
+            :param latents_name: The name of the latents tensor to get.
+            """
+            return services.latents.get(latents_name)
+
+        self.save = save
+        self.get = get
+
+
+class ConditioningInterface:
+    def __init__(
+        self,
+        services: InvocationServices,
+        context_data: InvocationContextData,
+    ) -> None:
+        # TODO(psyche): We are (ab)using the latents storage service as a general pickle storage
+        # service, but it is typed to work with Tensors only. We have to fudge the types here.
```

Comment:
This may be a good place to break it out into separate storage services

Reply:
Turns out we can use `ItemStorageABC` for this with a super tiny change it it. I've implemented this with support for custom save and load functions (defaulting to `torch.save` and `torch.load`).

Reply:
I've split the services up